src/main/haskell/language-kore/src/Data/Kore/Proof/Proof.hs

```diff
@@ -0,0 +1,421 @@
+
+type Path = [Int]
+type Proof = Prop Term LargeRule
+
+instance (Hashable formula, Hashable (rules subproof), Hashable assumption)
+  => Hashable (PropF formula rules assumption subproof) where
+  hashWithSalt s (ByF a b c) =
+    s `hashWithSalt` a `hashWithSalt` b `hashWithSalt` S.toList c
+
+instance Hashable Proof where
+    hashWithSalt s (By a b c) =
+        s `hashWithSalt` a `hashWithSalt` b `hashWithSalt` S.toList c
+
+data LargeRule subproof =
+ -- | a |- a
+   Assumption Term
+ -- | From a |- b deduce |- a -> b
+ | Discharge Term subproof
+ -- | From |- A[x] deduce |- Forall x. A[x]
+ | Abstract Var subproof
+ -- | EVIL SHORTCUT! This bypasses the functionality check.
+ -- Should move to using FunctionalSubst when you can.
+ | ForallElim Term subproof
+ -- | From |- a , |- b deduce |- a /\ b
+ | AndIntro subproof subproof
+ -- | From |- a /\ b deduce |- a
+ | AndElimL subproof
+ -- | From |- a /\ b deduce |- b
+ | AndElimR subproof
+ -- | a |- a \/ b
+ -- OrIntro a b
+ | OrIntroL subproof Term
+ -- | b |- a \/ b
+ -- OrIntro a b
+ | OrIntroR Term     subproof
+ -- | OrElim (a \/ b) (C assuming a) (C assuming b)
+ | OrElim subproof subproof subproof
+ -- | |- T
+ | TopIntro
+ -- | From |- A[x] deduce |- (E x. A[x])
+ | ExistsIntro Var Term subproof
+ -- | ExistsElim (E x. p[x]) (C assuming p[y])
+ | ExistsElim subproof Var Term subproof
+ -- | From |- a, |- a -> b deduce |- b
+ -- ModusPonens a b
+ | ModusPonens subproof subproof
+ -- (\forall x. phi) -> (\exists y. phi' = y) -> phi[phi'/x]
+ -- FunctionalSubst x phi y phi'
+ | FunctionalSubst Var Term Var Term
+ -- | \exists y . x = y
+ -- FunctionalVar x y
+ | FunctionalVar Var Var
+ | EqualityIntro Term
+ -- | Path points to the _subtree_ of phi in which the substitution
+ -- is to be applied at every possible point.
+ -- This is technically less flexible than specifying every position of "x"
+ -- But it's good enough for all practical purposes.
+ -- phi_1 = phi_2 /\ phi[phi_1/x] -> phi[phi_2/x]
+ -- EqualityElim phi_1 phi_2 phi [path]
+ | EqualityElim Term Term Term Path
+ -- | NOTE: Should probably rewrite axiom 8 to \forall x. x \in phi = \phi
+ -- This should be exactly equivalent, and it fits the other axioms better.
+ -- (\forall x . x \in phi) = phi
+ | MembershipForall Var Term
+ -- | x \in y = (x = y)
+ | MembershipEq Var Var
+ -- | x \in \not phi = \not (x \in phi)
+ | MembershipNot Var Term
+ -- | (x \in phi_1 /\ phi_2) = (x \in phi_1) /\ (x \in phi_2)
+ | MembershipAnd Var Term Term
+ -- | (x \in exists y . phi) = exists y . (x \in phi)
+ | MembershipExists Var Var Term
+ -- | x \in \sigma(phi_1,...,phi_i,...,phi_n)
+ -- =
+ -- \exists y . (y \in \phi_i /\ x \in \sigma(phi_1,...,y,...,phi_n))
+ -- MembershipCong x y i (\sigma(...))
+ | MembershipCong Var Var Int Term
+ deriving(Show, Functor, Foldable, Generic, Traversable)
+
+
+instance Hashable subproof => Hashable (LargeRule subproof)
+
+instance Pretty a => Pretty (LargeRule a) where
```

Comment:
Elsewhere, `Pretty` is used to print valid Kore, but this is not. Should I reinstate the separate `Unparse` class?

Reply:
I mean, at the moment, `Unparse` is defined as a class alias so that your proof types have `Unparse` instances. That's not right! But it's also my problem, not yours; I'm just remarking on it here.